lagom/container.py

```diff
@@ -408,6 +413,15 @@ def _infer_dependencies(
         }
         return {key: dep for (key, dep) in sub_deps.items() if dep is not None}
 
+    def _get_spec_and_arg_index(self, func: Callable[..., X]) -> Tuple[bool, int]:
+        if isinstance(func, FunctionType):
+            spec = self._reflector.get_function_spec(func)
+            arg_index = 0
+        else:
+            spec = self._reflector.get_function_spec(func.__init__)
+            arg_index = 1
```

Comment:
So just to check I've understood this correctly: The `arg_index` is needed here so that we don't supply `self`? I wonder if there's a way we could alter the spec in some way instead so we didn't need this :thinking:

Reply:
Yes that's correct. I didn't want to mess with the reflection code so this seemed like the simplest and most targeted approach, but removing `self` from `reflection.reflect` also works:

```python
def reflect(func: Callable) -> FunctionSpec:
    """
    Extension to inspect.getfullargspec with a little more.
    :param func:
    :return:
    """
    spec = inspect.getfullargspec(func)

    if "self" in spec.args:
        spec.args.remove("self")
    ...
```

This does cause one of the other tests to fail:

```python
def test_all_reflection_used_is_described(container: Container):
    container.resolve(AnotherThing)
    assert container.reflection_cache_overview == {
        "AnotherThing.__init__": "(?, AThing)",
        "AThing.__init__": "(?)",
    }
```

But that's simple enough to adjust the test to not expect the `?` for the now missing `self` param.

If you want me to go with this intead, I'm happy to make the change.

Reply:
How about changing `_get_spec_and_arg_index` then to something like:

```python
def _get_spec_without_self(self, func_or_class):
       spec = self._reflector.get_function_spec(func_or_class)
        if isinstance(func_or_class, FunctionType):
            return spec
        else:
            spec.args.remove("self")
            return spec
```

that way we keep the common reflection logic unchanged and just alter it where we know it needs altering?

Reply:
Indeed. We still need to check if self exists because of the caching in `_reflector`, but that works. I've made the change in the lastest updates